### ISSUE TITLE
S/MIME support

### ISSUE THREAD

**Issue by helmo** [k9mail/k-9#1003]

**Comment by sedrubal** [comment]
There was a mobile application development project at our university and one project tried to implement this: https://github.com/FAU-Inf2/SMile

For me this looks like a big chaos but maybe you can use some parts... ;)

**Comment by philipwhiuk** [comment]
We can take some lessons from it, but:

* The project layout is not how we'll want it. They have a separate git repository for the SMIME app which is referenced by SMile as a project dependency. What we want is to do similarly to the OpenPGP system and to create a SMIME library inside the same GitHub repository which dispatches and receives Intents from an SMIME app. Otherwise it gets quite difficult to build the app.

* They've changed a lot of files which are not related to SMIME. For example they've added the ability to logout from an IMAP account by the look of it, added some form of reminder functionality, added lots of text changes and so on. This makes it substantially more complex to merge in just the SMIME changes. Even cherry-picking commits will be a pain because the commits aren't obvious what they do.

Similarly to the first OpenPGP implementation, we can probably learn lessons but direct code re-use is going to be problematic.

**Comment by philipwhiuk** [comment]
I've begun a branch in my local repository which assumes it will receive similar intents to that which OpenKeychain provides: https://github.com/philipwhiuk/k-9/commits/smime I will keep re-basing this on-top of master maintaining parity with any new relevant features that the PGP/MIME development adds.

I need to write a SMIME certificates storage app now to implement the other side so I can test decrypting a SMIME message (and test composition). It's possible I can fork https://github.com/FAU-Inf2/SMileCE for that. Otherwise it will take a while...

**Comment by cketti** [comment]
Right now the OpenKeychain project kind of owns the crypto provider interface and drives the development. Which is fine, because nobody else showed interest in integrating with K-9 Mail.

But really I want to have a crypto provider interface that is generic enough to be able to support both PGP/MIME and S/MIME. That is also part of the reason why I suggested using the openintents namespace back when this all started. My hope was this interface will be implemented by more than one app, and one day supports more than one crypto system.
The use of "openpgp" all over the place is a bit unfortunate. But I suspect the API needs some tweaking to be able to support S/MIME anyway. So transitioning to a more generic name shouldn't be a huge problem either.

/cc @Valodim @dschuermann

**Comment by adeorato** [comment]
I can recommend the library https://en.wikipedia.org/wiki/Bouncy_Castle_(cryptography) for this.  At work we use this API for Java and S/Mime and it was simple to integrate. It also supports PGP, but I never tried the PGP part.

**Comment by szepeviktor** [comment]
While you implement this a temporary word-around is [X509Tools](https://play.google.com/store/apps/details?id=at.rundquadrat.android.x509tools)

**Comment by trollkarlen** [comment]
Any progress or news about this ?

Did I understand this right that it was going to be handled by a external program like OpenKeychain for OenPGP by an intent ?
But then is this a separate program or in OpenKeychain ?

/T

**Comment by GTAXL** [comment]
What is the status on this? S/MIME is nearly supported in every e-mail client. It's been around just as long as PGP and is actually easier to use than PGP and you can verify origin from a trusted CA.

**Comment by errhammr** [comment]
Hello everyone! I'm curious, what's the current status of the S/MIME integration?
Are there still plans to ship this feature, do you need help with something specific, or is it just not going to happen?

**Comment by szepeviktor** [comment]
In 3 years you grew a long 🧔

**Comment by dhdurgee** [comment]
I need an Android email app with S/MIME support.  I see this issue is still open, but that a milestone for it was removed last month.  Is there a version with S/MIME under development at this point?  If so, can I test it for you?  I am running a Motorola Moto E with Android 10 on it that would be my test environment.

**Comment by ShogunWeb** [comment]
Also interested in having an open source Android mail client that supports S/MIME. It seems this protocol is well integrated in most desktop clients, unlike PGP, which makes it the preferred choice for companies. As I need a mobile client mail for my pro email (having a S/MIME certificate), I would like to know if I can wait for a solution using my favorite K-9 client, or should I switch to another open source alternative ? 

In case S/MIME will not be supported in the short terme in K-9, would you recommend an open source alternative ? (FairMail looks like a very good option)

**Comment by witchent** [comment]
> Also interested in having an open source Android mail client that supports S/MIME. It seems this protocol is well integrated in most desktop clients, unlike PGP, which makes it the preferred choice for companies. As I need a mobile client mail for my pro email (having a S/MIME certificate), I would like to know if I can wait for a solution using my favorite K-9 client, or should I switch to another open source alternative ?
> 
> In case S/MIME will not be supported in the short terme in K-9, would you recommend an open source alternative ? (FairMail looks like a very good option)

I also had to switch because I needed S/MIME support, and FairMail works fine, though I'd still prefer K-9 very much if only for S/MIME..

**Comment by rezazarchi** [comment]
I researched this feature in some K-9 forks like [SMile](https://github.com/FAU-Inf2/SMile) and [K-2](https://github.com/philipwhiuk/k2/tree/smime). But they didn't actually implement this feature completely. 
For example, in Smile sign process, they didn't pass the email body to [SMileCE](https://github.com/FAU-Inf2/SMileCE) service, and it returns error with null pointer exception. In the [SMileCE](https://github.com/FAU-Inf2/SMileCE) project, MimeBodyPart object is from javax mail library, which is different from our MimeBodyPart in K-9

**Comment by ShogunWeb** [comment]
Thanks @rezazarchi for the info. I did not know these forks, but anyway, from what I read, S/MIME is not fully implemented so it won't save me :-)

**Comment by rezazarchi** [comment]
I'm working on S/mime in the K-9 project. but I have some challenges with converting K9 MimeMessage to Javax MimeMessage objects and converting Javax to K9. 
The bouncy castle needs that to sign messages. 
I created a signed message successfully, but for reading and validating signator, I had challenge with converting saved local message to Javax mail. I don't know how to convert **K9 LocalMessage** to **Javax MimeMessage** properly. 
Could you please help me with that? @cketti 

Part of my codes for converting is here. I will push my whole code when It is completed.
https://gist.github.com/rezazarchi/8923ef38433677b77943195624cb6f71

**Comment by cketti** [comment]
@rezazarchi: I assume JavaMail (or I guess Jakarta Mail now) can read and write raw messages. Probably easiest to go that route.

If you want to get this merged into K-9 Mail, please find another way. I don't want to include another (rather large) mail library like Jakarta Mail. I'm sure BouncyCastle has more low level functionality to encrypt/decrypt and sign/verify S/MIME data.

**Comment by rezazarchi** [comment]
So how can I read raw message from LocalMessage?

**Comment by cketti** [comment]
Use `Body.writeTo(OutputStream)`. See e.g. `MessageCryptoHelper.getDataSourceForEncryptedOrInlineData()`.

**Comment by schokotets** [comment]
Fellow developers, I've got a few questions regarding S/MIME support:
- Where does development for it take place?
- How's progress?
- Is it possible to support the development of S/MIME with targeted donations?
- Will it land on @thundernest's Android Roadmap?

**Comment by AlexanderKaschta** [comment]
The comments in this thread about the code design are eight years old and I've wanted to check, if they are still valid:

- Is it still intended to implement S/MIME in such a way, that it uses a similar crypto interface to the PGP implementation (including an external application for key management)? Are there any arguments against directly integrating the complete logic (key import, key management, viewing keys, encryption and signing) into thunderbird-android, to have a similar experience to the desktop app?

**Comment by user4andre** [comment]
I would be pleased if the support of S/MINE was not postponed again and again. 

The Fairmail app (https://github.com/M66B/FairEmail) supports S/MINE.
There is no additional APP for the S/MINE certificates, but according to the FAQ, Android internal storage is used.

from the FAQ page:

S/MIME
Encrypting a message requires the public key(s) of the recipient(s). Signing a message requires your private key.

Private keys are stored by Android and can be imported via the Android advanced security settings (Encryption & credentials, Install a certificate, VPN & app user certificate). There is a shortcut (button) for this in the encryption settings for Android version 10 and before. Android will ask you to set a PIN, pattern, or password if you didn't before. 

(...) shortened!

**Comment by PhilippSchlesinger** [comment]
There is a corresponding feature [request](https://connect.mozilla.org/t5/ideas/s-mime-support-for-k-9-mail-android/idi-p/61756) on Mozilla Connect.

**Comment by PhilippSchlesinger** [comment]
@kewisch This feature request has received a lot of thumbs over time here and in the [request] (https://connect.mozilla.org/t5/ideas/s-mime-support-for-k-9-mail-android/idi-p/61756) on Mozilla Connect.
How will it be reclassified within https://github.com/orgs/thunderbird/projects/19?

**Comment by kewisch** [comment]
Unfortunately it won't change the roadmap for this year, but when it comes to developing next year's roadmap the votes and general direction will certainly play a role. https://github.com/thunderbird/thunderbird-android/issues/1003#issuecomment-1699890147 is still correct today.

**Comment by trollkarlen** [comment]
If you wait long enough there will be a new standard to not implement so
this will be voided 😃

//R

______________________________________________
Robert Marklund

Mobile: +46 (0)70 213 22 76
E-mail: ***@***.***
Chat:   ***@***.***
______________________________________________

**Comment by seoMattH** [comment]
> Unfortunately it won't change the roadmap for this year, but when it comes to developing next year's roadmap the votes and general direction will certainly play a role. [#1003 (comment)](https://github.com/thunderbird/thunderbird-android/issues/1003#issuecomment-1699890147) is still correct today.

I mean does this mean that you guys just aren't ever going to do it?  I'm curious because waiting endlessly doesn't seem to be a good solution.  I would prefer you guys just come out and say.  "Nah we have no intention of doing it"

**Comment by MANICX100** [comment]
> > Unfortunately it won't change the roadmap for this year, but when it comes to developing next year's roadmap the votes and general direction will certainly play a role. [#1003 (comment)](https://github.com/thunderbird/thunderbird-android/issues/1003#issuecomment-1699890147) is still correct today. [@kewisch](https://github.com/kewisch)
> 
> I mean does this mean that you guys just aren't ever going to do it? I'm curious because waiting endlessly doesn't seem to be a good solution. I would prefer you guys just come out and say. "Nah we have no intention of doing it"

I understand the frustration, personally this is the one reason why I use Fair Email instead.

I'm sure there are people working on thunderbird who would also like to see this feature, but peoples job follow processes for good and ill.

2026 is therefore the earliest this will be looked at, and being vote dependant it is very possible it never will be.

I work as a Microsoft Power Platform Developer and I find it eternally annoying that 'features' and even 'bug-fixes' in key products get forced to compete for 'community votes' which in my opinion is limited data based on a minority user base leading to shall we say stunted usability growth.

 But thunderbird is not closed source, someone could implement it and create a pull request. I am not saying it should be you, I am not saying it would be me (I'm not good enough), but that might be the way.

**Comment by kewisch** [comment]
Community votes is one source for the roadmap, but not the only one. Increasing the use of encryption is a goal the Thunderbird Councils is very interested in, so it might well be part of next year's roadmap even if it is not the top voted community feature.

I can also understand the frustration, but flat out saying we'll never do it wouldn't be accurate either. What if we do? Then I walked back on my "promise"? I don't feel it would be useful to make a statement on this right now. If someone feels they want to make an investment into S/MIME support we can certainly work with them to find out how to best get it into the app.

**Comment by PhilippSchlesinger** [comment]
@kewisch Thank you for the valuable insights you provided.

Maybe it would be good to post a request for help on the [Thunderbird e2ee mailing list](https://thunderbird.topicbox.com/groups/e2ee)?

Do you think it would be possible to integrate S/MIME support into [OpenKeychain](https://github.com/open-keychain)?
@user4andre stated above that the Fairmail app (https://github.com/M66B/FairEmail) supports S/MIME. Possibly there is a way to work with @M66B on improving and making S/MIME support available as a library?

**Comment by wiktor-k** [comment]
> Do you think it would be possible to integrate S/MIME support into [OpenKeychain](https://github.com/open-keychain)?

FWIW OpenKeychain is no longer developed (see "WARNING: This software is no longer actively developed." in [the README](https://github.com/open-keychain/open-keychain/blob/master/README.md)). Even if it were, it's deeply tied to PGP so S/MIME support would be unlikely.

Another thing is that Android has [certificate store](https://stackoverflow.com/questions/57661796/android-app-storing-client-certificates-in-a-secure-and-proper-way) built-in, so no need for a certificate manager for S/MIME.

(Just to be in the clear: I'm not saying this feature is simple to implement. Only clarifying some points).

**Comment by poettig** [comment]
> Another thing is that Android has [certificate store](https://stackoverflow.com/questions/57661796/android-app-storing-client-certificates-in-a-secure-and-proper-way) built-in, so no need for a certificate manager for S/MIME.

That's true and in use by the already cited FairEmail. Unfortunately, Android only includes root certificates for TLS certificates, not for S/MIME. As most globally trusted CAs have multi-purpose roots, this works for many certificates, but there are CAs using separate roots for S/MIME and TLS. This breaks verification of signatures by default on FairEmail, but it works in Thunderbird and Outlook (Desktop), as Mozilla (https://wiki.mozilla.org/CA/Included_Certificates) and Microsoft have special trust programs.

A solution including the Mozilla Root Store in the app should therefore be highly preferred in my opinion.

**Comment by wiktor-k** [comment]
> A solution including the Mozilla Root Store in the app should therefore be highly preferred in my opinion.

Yes, I was thinking about using certificate store for user certificates ("app user certificate") to keep the e-mail client from handling software keys (AFAIK Android can use hardware storage for private keys).

Thanks for the wider context with S/MIME CAs.

**Comment by poettig** [comment]
> Thanks for the wider context with S/MIME CAs.

If an example helps on any point in the future, we currently have the described problem with FairEmail because our certificate provider [HARICA](https://repo.harica.gr/rep_dyn.php) has separate roots for TLS and S/MIME.

**Comment by MANICX100** [comment]
> > Do you think it would be possible to integrate S/MIME support into [OpenKeychain](https://github.com/open-keychain)?
> 
> FWIW OpenKeychain is no longer developed (see "WARNING: This software is no longer actively developed." in [the README](https://github.com/open-keychain/open-keychain/blob/master/README.md)). Even if it were, it's deeply tied to PGP so S/MIME support would be unlikely.
> 
> Another thing is that Android has [certificate store](https://stackoverflow.com/questions/57661796/android-app-storing-client-certificates-in-a-secure-and-proper-way) built-in, so no need for a certificate manager for S/MIME.
> 
> (Just to be in the clear: I'm not saying this feature is simple to implement. Only clarifying some points).

Desktop Thunderbird however supports its own certificate store separate to the underlying OS e.g. windows/macos both have their own certificate store.

**Comment by caguiar** [comment]
S/MIME (certificate) signing and encryption, as per Thunderbird Desktop, would be a very welcome feature.
I have installed S/MIME email certificate on my Linux (Thunderbird Desktop) and on my Mac (Mail), now all I need is to be able to install it on my Thunderbird Mobile.

**Comment by dubiago** [comment]
I second caguiar's sentiments. I have Thunderbird set up for both S/MIME and PGP on desktop, and that inconsistency is going to get quite untenable over time.

**Comment by seoMattH** [comment]
Without question the lack of S/MIME on mobile is really annoying.  Seems like an easy fix to make and would make a massive difference for security conscious users.

**Comment by dubiago** [comment]
To be fair, they're currently using OpenKeychain for key management for PGP; they'd have to shift to using the native KeyStore/KeyChain for storing S/MIME certs. Not a tremendously huge lift, but I'm sure they have other priorities. But, it looks like this has been open for a _very_ long time. I do hope it's addressed soon.

**Comment by ThomasKramps** [comment]
May it be a good idea to place this in your backlog (https://github.com/orgs/thunderbird/projects/19/views/4) now?

Just to be sure it will be discussed for future roadmaps.
